### DISCUSSION TITLE
Development environment

### DISCUSSION THREAD

**Discussion by BenasB** [vu-vibedosa/apsitvarkom-be#40] [Announcements]
## Continuous delivery/deployment
We're already using CI for our PRs, so the next logical step is to embrace [CICD](https://www.redhat.com/en/topics/devops/what-is-ci-cd) as a whole – implement the CD part. For the time being, only a `dev`elopment environment should suffice our needs.

## What does that mean for developers?
When you push changes to the `dev` branch, an automated system will pick it up, rebuild our projects and deploy them – we'll have a dedicated environment showing the current development state of our application. You will be able to go to some URL and see the application (as if it was "released" but just for development purposes). This is a common practice. As developers, we'll be able to see our progress on the application incrementally without needing to pull new changes and run them locally.

## The plan
I have spent a significant amount of time researching various options and decided on the plan:
- FE and BE projects are [containerized](https://www.youtube.com/watch?v=0qotVMX-J5s&ab_channel=IBMTechnology)
- All infrastructure is defined as [Infrastructure as Code](https://en.wikipedia.org/wiki/Infrastructure_as_code) using [Terraform](https://www.terraform.io/) – a new repo `apsitvarkom-infra` created
- Use [Google Cloud Platform (GCP)](https://cloud.google.com/) and utilize its [free tier](https://cloud.google.com/free)
  - Docker images stored in [Artifact Registry](https://cloud.google.com/artifact-registry)
  - The actual instances of FE and BE are ran in [Cloud Run](https://cloud.google.com/run)
  - [Cloud Storage](https://cloud.google.com/storage) will store Terraform state and other files if we need – most probably we can even store images there.
  - [Cloud SQL](https://cloud.google.com/sql) does not have a free tier :/ but we can have one `e2-micro` [Compute Engine](https://cloud.google.com/compute) VM and run whatever SQL server we like, so we'll be fine.
- [GitHub Actions](https://github.com/features/actions) builds our images and pushes them to Artifact Registry

*P.S. I will handle this part, since it's pretty advanced stuff (for a uni project)*

**Comment by 6ar8nas** [comment]
🚀🚀🚀

**Comment by BenasB** [comment]
Back end is is being deployed, go to the main repo page and on the right you'll see:

![image](https://user-images.githubusercontent.com/29711974/194427633-961511fc-6fb3-414a-9f9d-48842b236590.png)

Then you can click "View deployment"
![image](https://user-images.githubusercontent.com/29711974/194427668-6e1e3b33-31af-4b2b-ba37-dfe7133bf660.png)

It will take you to some link, which will not show anything (because our API does not have an action on route `/`) but add `/swagger` at the end and you will be taken to the usual swagger page 🎉

**Comment by BenasB** [comment]
Made some changes to the way BE is deployed and also added FE deployment:
![image](https://user-images.githubusercontent.com/29711974/194776355-a10b2a99-b454-47d3-943b-e72b10f015ff.png)

So if you go to the DEV FE, it will load our front end that is calling our DEV BE – now we have a complete development environment

**Reply by 6ar8nas** [reply]
Nice seeing this available already. Great job.

**Comment by BenasB** [comment]
Update: `dev` is now `staging` see https://github.com/vu-vibedosa/apsitvarkom-infra/issues/14 for more information.
